### PR TITLE
Introduce Authentication subsystem facade methods.

### LINKED ISSUE #1408: Enhance new authentication subsystem API
Namely, encapsulate common steps of handling authentication transaction (with various objects interaction) into an `AuthenticationSystemSupport` façade. Introduce 4 facade methods:

```java
AuthenticationContextBuilder handleInitialAuthenticationTransaction(Credential c);

AuthenticationContext handleFinalizedAuthenticatioAttempt(Credential c, Service s);

void handleAuthenticationTransaction(AuthenticationContextBuilder b, Credential c);

AuthenticationContext finalizeAuthenticationAttempt(AuthenticationContextBuilder b, Service s);

```

so this repeated pattern (assuming multiple authentication txs):

```java
final AuthenticationContextBuilder builder = new DefaultAuthenticationContextBuilder(
                    this.authenticationSystemSupport.getPrincipalElectionStrategy());
            final AuthenticationTransaction transaction =
                    AuthenticationTransaction.wrap(credential);
            this.authenticationSystemSupport.getAuthenticationTransactionManager()
                    .handle(transaction,  builder);

         ....

         final AuthenticationTransaction transaction2 =
                    AuthenticationTransaction.wrap(credential2);
            this.authenticationSystemSupport.getAuthenticationTransactionManager()
                    .handle(transaction2,  builder);
      
      ...

       final AuthenticationTransaction transaction3 =
                    AuthenticationTransaction.wrap(credential3);
            this.authenticationSystemSupport.getAuthenticationTransactionManager()
                    .handle(transaction3,  builder);
```

would become this:

```java
AuthenticationContextBuilder b = authenticationSystemSupport.handleInitialAuthenticationTransaction(cred1);
...

authenticationSystemSupport.handleAuthenticationTransaction(b, cred2);

...

authenticationSystemSupport.handleAuthenticationTransaction(b, cred3);

...
//and then

AuthenticationContext authCtx = authenticationSystemSupport.finalizeAuthenticationAttempt(b, service);

```

and this (single authentication transaction) pattern:

```java
final AuthenticationContextBuilder builder = new DefaultAuthenticationContextBuilder(
                    this.authenticationSystemSupport.getPrincipalElectionStrategy());
            final AuthenticationTransaction transaction =
                    AuthenticationTransaction.wrap(credential);
            this.authenticationSystemSupport.getAuthenticationTransactionManager()
                    .handle(transaction,  builder);
            final AuthenticationContext authenticationContext = builder.build(serviceTicket.getService());
```

would become this:

```java
AuthenticationContext authCtx = authenticationSystemSupport.handleFinalizedAuthenticationAttempt(cred, service)
```

### PR DESCRIPTION
**Pull request by dima767**
Closes #1408

### CODE REVIEW

**Comment by mmoayyed** [comment]
This is pretty good. Thanks! I am however pretty sure that there are more places that could use the new API. Did you scan the codebase for all places that use the new support class? I bet there are more test cases that use utilize the "old" model.

**Comment by dima767** [comment]
I did the pretty thorough scan via IDEA. I think I got it all, but if for some reason something is missed that could be refactored later on "on the fly" ;-)

**Comment by mmoayyed** [comment]
Kewl.

**Comment by mmoayyed** [comment]
:+1:

File: cas-server-core-api-authentication/src/main/java/org/jasig/cas/authentication/AuthenticationSystemSupport.java [hunk 1]
```diff
@@ -23,4 +31,53 @@
...
+     * Initiate potential multi-transaction authentication attempt by handling the initial authentication transaction.
+     *
+     * @param credential a credential for the initial authentication transaction.
+     *
+     * @return authentication context builder used to accumulate authentication transactions in this authentication attempt.
+     *
+     * @throws AuthenticationException exception to indicate authentication processing failure.
+     * @since 4.3.0
+     */
+    AuthenticationContextBuilder handleInitialAuthenticationTransaction(Credential... credential) throws AuthenticationException;
```
Comment (leleuj) [inline cas-server-core-api-authentication/src/main/java/org/jasig/cas/authentication/AuthenticationSystemSupport.java]: `initializeAuthenticationTransaction` instead?

Comment (leleuj) [inline cas-server-core-api-authentication/src/main/java/org/jasig/cas/authentication/AuthenticationSystemSupport.java]: No, you're right, it's really about handling the initial authentication transaction.

**Comment by leleuj** [comment]
I like the idea, I'm not fully convinced by the names. Could we have something more explicit like `startTransactions`, `addTransaction` and `endTransactions`?

**Comment by dima767** [comment]
But I did carefully think about the current names. We have 2 concepts here,
really - an authentication transaction - a single interaction with an
authentication source with given credentials and authentication attempt - a
summary of all the possible auth transactions in order to gain access,
create web SSO session, etc. Hence the very carefully chosen names to
expose an intention revealing interface.

D.

**Comment by leleuj** [comment]
I was not respectfully saying that names have not been properly devised. The `attempt` concept is something which has never been used before and which is not really clear not me. When I read `attempt`, I think: `unfinished` and `(maybe failed`.
Though, the idea here is when we have played all transactions, to finish the authentication process and generate an authentication context, right?

**Comment by dima767** [comment]
Yes, "attempt" is somewhat new concept here, but nevertheless we just
saying that someone is attempting to authenticate. A "finalized attempt" is
successful when all the necessary authentication transactions are executed
and final authentication context is built.

So if "attempt" does not represent a clear concept here, may be we could
refer to it (in the facade method names) as "authentication result". Or
even better - rename AuthenticationContext to AuthenticationResult!

**Comment by leleuj** [comment]
I like `AuthenticationResult`.

**Comment by dima767** [comment]
Misagh, let's rename AuthenticationContext to AuthenticationResult

**Comment by mmoayyed** [comment]
Sure. Questions: 

Is this something:

1. We should do in this pull?
2. I should do, in this pull?
3. I should do in a separate pull?

**Comment by mmoayyed** [comment]
P.S: The name sounds good to me, if you want to make changes in the same pull, etc.

**Comment by dima767** [comment]
How about this - merge this pull and I'll do the refactoring from the
merged master and make a separate PR. Will be cleaner. Sounds good?

**Comment by mmoayyed** [comment]
Copy that. @leleuj OK with you?

**Comment by mmoayyed** [comment]
PS renamed the pull description to use the keyword "closes" as that would automatically close the issue when the PR is merged.

**Comment by leleuj** [comment]
+1

**Comment by dima767** [comment]
And so with the refactored `AuthenticationResult` the new facade API would look like this (which IMHO very clear and intention revealing):

```java
AuthenticationResultBuilder handleInitialAuthenticationTransaction(Credential... credential)

    AuthenticationResultBuilder handleAuthenticationTransaction(AuthenticationResultBuilder authenticationResultBuilder, Credential... credential)

    AuthenticationResult finalizeAllAuthenticationTransactions(AuthenticationResultBuilder authenticationResultBuilder, Service service)

    AuthenticationResult handleAndFinalizeSingleAuthenticationTransaction(Service service, Credential... credential)
```

**Comment by dima767** [comment]
Approx ETA for the new PR on this  -some time on (01/12/2016 or 01/13/2016)

**Comment by leleuj** [comment]
I agree

**Pull request merged**